dbms/src/Functions/FunctionsJson.h

```diff
@@ -530,11 +612,671 @@ class FunctionsJsonArray : public IFunction
+                    data_to,
+                    offsets_to,
+                    input_source,
+                    input_tidb_tp->tp(),
+                    input_tidb_tp->flen(),
+                    block.rows());
+            }
+            else
+            {
+                doExecuteForBinary<false>(
+                    data_to,
+                    offsets_to,
+                    input_source,
+                    input_tidb_tp->tp(),
+                    input_tidb_tp->flen(),
+                    block.rows());
+            }
+        }
+        else if ((unlikely(output_tidb_tp == nullptr)) || hasParseToJSONFlag(*output_tidb_tp))
+        {
+            if (from.column->isColumnNullable())
+            {
+                const auto & column_nullable = static_cast<const ColumnNullable &>(*from.column);
+                doExecuteForParsingJson<true>(
+                    data_to,
+                    offsets_to,
+                    input_source,
+                    column_nullable.getNullMapData(),
+                    block.rows());
+            }
+            else
+            {
+                auto tmp_null_map = ColumnUInt8::create(0, 0);
+                doExecuteForParsingJson<false>(
+                    data_to,
+                    offsets_to,
+                    input_source,
+                    tmp_null_map->getData(),
+                    block.rows());
+            }
+        }
+        else
+        {
+            doExecuteForOthers(data_to, offsets_to, input_source, block.rows());
+        }
+
+        if (from.column->isColumnNullable())
+        {
+            auto null_map = static_cast<const ColumnNullable &>(*from.column).getNullMapColumnPtr();
+            block.getByPosition(result).column = ColumnNullable::create(std::move(col_to), std::move(null_map));
+        }
+        else
+        {
+            block.getByPosition(result).column = std::move(col_to);
+        }
+    }
+
+private:
+    template <bool is_binary_str>
+    static void doExecuteForBinary(
+        ColumnString::Chars_t & data_to,
+        ColumnString::Offsets & offsets_to,
+        const std::unique_ptr<IStringSource> & data_from,
+        UInt8 from_type_code,
+        Int32 flen,
+        size_t size)
+    {
+        size_t reserve_size = (size * (1 + 1)) + data_from->getSizeForReserve();
+        JsonBinary::JsonBinaryWriteBuffer write_buffer(data_to, reserve_size);
+        for (size_t i = 0; i < size; ++i)
+        {
+            const auto & slice = data_from->getWhole();
+            if constexpr (is_binary_str)
+            {
+                if (unlikely(flen <= 0))
+                {
+                    JsonBinary::appendOpaque(
+                        write_buffer,
+                        JsonBinary::Opaque{from_type_code, StringRef{slice.data, slice.size}});
+                }
+                else
+                {
+                    auto size_t_flen = static_cast<size_t>(flen);
```

Comment:
The `reserve_size` for this case is `size*(1 + 1 + size_t_flen)`?

Reply:
yes, updated.